ast/ArgParsing.cc

```diff
@@ -40,41 +41,64 @@ ParsedArg ArgParsing::parseArg(core::Context ctx, unique_ptr<ast::Reference> arg
     return parsedArg;
 }
 
-vector<ParsedArg> ArgParsing::parseArgs(core::Context ctx, ast::MethodDef::ARGS_store &args) {
+unique_ptr<ast::Expression> getDefaultValue(unique_ptr<ast::Expression> arg) {
+    auto *refExp = ast::cast_tree<ast::Reference>(arg.get());
+    if (!refExp) {
+        Exception::raise("Must be a reference!");
+    }
+    unique_ptr<ast::Expression> default_;
+    typecase(
+        refExp, [&](ast::UnresolvedIdent *nm) { Exception::raise("Unexpected unresolved name in arg!"); },
+        [&](ast::RestArg *rest) { default_ = getDefaultValue(move(rest->expr)); },
+        [&](ast::KeywordArg *kw) { default_ = getDefaultValue(move(kw->expr)); },
+        [&](ast::OptionalArg *opt) { default_ = move(opt->default_); },
+        [&](ast::BlockArg *blk) { default_ = getDefaultValue(move(blk->expr)); },
+        [&](ast::ShadowArg *shadow) { default_ = getDefaultValue(move(shadow->expr)); },
+        [&](ast::Local *local) {
+            // No default.
+        });
+    ENFORCE(default_ != nullptr);
+    return default_;
+}
+
+} // namespace
+
+vector<ParsedArg> ArgParsing::parseArgs(const ast::MethodDef::ARGS_store &args) {
     vector<ParsedArg> parsedArgs;
     for (auto &arg : args) {
         auto *refExp = ast::cast_tree<ast::Reference>(arg.get());
         if (!refExp) {
             Exception::raise("Must be a reference!");
         }
         unique_ptr<ast::Reference> refExpImpl(refExp);
-        arg.release();
-        parsedArgs.emplace_back(parseArg(ctx, move(refExpImpl)));
+        parsedArgs.emplace_back(parseArg(refExpImpl));
+        // Don't free the pointer; it's still owned by `args`.
+        refExpImpl.release();
```

Comment:
This creates opportunity for memory leak if `parseArg` throws.
Can we get it to state it used to be(where there is no calls between ownership move and release)?

Reply:
Actually, I don't even need a `unique_ptr` here; I can pass a reference to `parseArg` since it's `const` now. :D